### PR TITLE
Optimized sort

### PR DESCRIPTION
**Pull request by jonathanpoelen**
```cpp
using list = brigand::sort<brigand::range<int, 0, SIZE>>;
```

gcc-5.4.0:

Size = |      10     |      100     |      200     |      300     |      500    
------ | ----------- | ------------ | ------------ | ------------ | ------------
before | 0.07s - 26M | 0.42s - 148M | 1.19s - 344M | 2.28s - 609M | 4.00s - 1011M
after  | 0.05s - 19M | 0.16s -  45M | 0.38s -  94M | 0.66s - 152M | 1.11s - 245M
gain   |  28%  - 24% |  61%  -  69% |  68%  -  72% |  71%  -  75% |  72%  -  75%

clang-3.8.0:

Size = |      10     |     100     |     200     |      300     |      500    
------ | ----------- | ----------- | ----------- | ------------ | ------------
before | 0.09s - 40M | 0.26s - 54M | 0.66s - 95M | 1.38s - 180M | 2.49s - 321M
after  | 0.09s - 40M | 0.21s - 50M | 0.48s - 73M | 0.78s - 101M | 1.27s - 145M
gain   |   0%  -  0% |  19%  -  6% |  27%  - 23% |  43%  -  43% |  48%  -  54%

```cpp
using list = brigand::sort<brigand::reverse_range<int, SIZE, 0>>;
```

gcc-5.4.0:

Size = |      10     |     100     |      200     |      300     |      500    
------ | ----------- | ----------- | ------------ | ------------ | ------------
before | 0.05s - 21M | 0.25s - 83M | 0.46s - 152M | 0.70s - 223M | 1.37s - 375M
after  | 0.05s - 19M | 0.14s - 40M | 0.26s -  65M | 0.39s -  93M | 0.71s - 161M
gain   |   0%  -  7% |  43%  - 51% |  43%  -  56% |  44%  -  58% |  48%  -  56%

clang-3.8.0:

Size = |      10     |     100     |     200     |     300     |      500    
------ | ----------- | ----------- | ----------- | ----------- | ------------
before | 0.08s - 40M | 0.19s - 48M | 0.34s - 63M | 0.52s - 81M | 0.98s - 129M
after  | 0.09s - 40M | 0.19s - 49M | 0.32s - 60M | 0.47s - 73M | 0.84s - 108M
gain   | -12%  - -1% |   0%  -  0% |   5%  -  4% |   9%  -  8% |  14%  -  16%

```cpp
using list = brigand::sort<
  brigand::transform<
    brigand::range<int, 0, SIZE>,
    brigand::bind<brigand::modulo, brigand::_1, std::integral_constant<int, 40>>
  >
>;
```

gcc-5.4.0:

Size = |      10     |      100     |      200     |      300     |      500    
------ | ----------- | ------------ | ------------ | ------------ | ------------
before | 0.07s - 27M | 0.53s - 177M | 1.66s - 469M | 3.02s - 789M | 5.89s - 1406M
after  | 0.05s - 21M | 0.22s -  56M | 0.54s - 121M | 0.92s - 199M | 1.90s - 341M
gain   |  28%  - 23% |  58%  -  68% |  67%  -  74% |  69%  -  74% |  67%  -  75%

clang-3.8.0:

Size = |      10     |     100     |      200     |      300     |      500    
------ | ----------- | ----------- | ------------ | ------------ | ------------
before | 0.10s - 41M | 0.33s - 60M | 0.96s - 127M | 1.76s - 217M | 3.83s - 449M
after  | 0.10s - 41M | 0.26s - 55M | 0.60s -  87M | 1.04s - 131M | 1.81s - 204M
gain   |   0%  - -1% |  21%  -  7% |  37%  -  31% |  40%  -  39% |  52%  -  54%

```cpp
using list = brigand::sort<brigand::append<
  brigand::range<int, SIZE/2, SIZE>,
  brigand::range<int, 0, SIZE/2>
>>;
```

gcc-5.4.0:

Size = |      10     |      100     |      200     |      300     |      500    
------ | ----------- | ------------ | ------------ | ------------ | ------------
before | 0.07s - 25M | 0.46s - 158M | 1.11s - 328M | 1.86s - 534M | 3.38s - 903M
after  | 0.05s - 19M | 0.17s -  45M | 0.33s -  82M | 0.63s - 139M | 1.13s - 244M
gain   |  28%  - 20% |  63%  -  71% |  70%  -  74% |  66%  -  73% |  66%  -  72%

clang-3.8.0:

Size = |      10     |     100     |     200     |      300     |      500    
------ | ----------- | ----------- | ----------- | ------------ | ------------
before | 0.09s - 40M | 0.28s - 55M | 0.61s - 86M | 1.13s - 138M | 2.02s - 245M
after  | 0.09s - 41M | 0.21s - 51M | 0.40s - 67M | 0.71s - 101M | 1.27s - 146M
gain   |   0%  -  0% |  25%  -  8% |  34%  - 22% |  37%  -  27% |  37%  -  40%

### CODE REVIEW

**Comment by porkybrain** [comment]
Nice job! I think we can optimize even further by using the "universal fast track approach". still 500 elements in just over a second is amazing!

File: brigand/algorithms/merge.hpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    template<class... R, class Comp>
+    struct merge_impl<list<R...>, list<>, list<>, Comp>
+    {
+      using type = list<R...>;
+    };
+  }
+
+  template<class Seq1, class Seq2, class Comp = less<_1,_2>>
+  using merge = append<clear<Seq1>, typename detail::merge_impl<clear<Seq1>, wrap<Seq1, list>, wrap<Seq2, list>, Comp>::type>;
```
Comment (porkybrain) [inline brigand/algorithms/merge.hpp]: don't you mean `detail::merge_impl<list<>, wrap<Seq1, list>, wrap<Seq2, list>, Comp>::type` here?

Comment (jonathanpoelen) [inline brigand/algorithms/merge.hpp]: The rogue, he escaped.

**Comment by jonathanpoelen** [comment]
> I think we can optimize even further by using the "universal fast track approach".

Probably. I tried some tricks, but without results. I think tried 1 or 2 other things.
Next step, less than a second with 500 elements? :D

**Comment by jonathanpoelen** [comment]
I just test again with the "universal fast track" by applying it directly on `detail::apply`:

```cpp
        template <template<typename,typename> class F, typename T, typename U>
        struct apply<F<_1,_2>, list<T, U>> : F<T, U>
        {
        };

        template <template<typename,typename> class F, typename T, typename U>
        struct apply<F<_2,_1>, list<T, U>> : F<U, T>
        {
        };
```

gcc: ~+20%
clang: ~+6%

**Comment by porkybrain** [comment]
Yes under one second, the race is on ;) for kvasir I need 1000 elements to not crash and not be more than the default max instantiation depth

**Comment by ldionne** [comment]
Is this sort stable?

**Comment by porkybrain** [comment]
I think #208 is the last bug. I have been hammering it.

**Comment by jonathanpoelen** [comment]
> Is this sort stable?

No, but it could.

In the meantime.
```cpp
template<class L, class Comp> using stable_sort
  = sort<L, not_<bind<apply, pin<Comp>, _2, _1>>>;
```

**Comment by edouarda** [comment]
@ldionne do you have any specific use case in mind for a stable sort? I'm curious.

**Comment by ldionne** [comment]
It's just that Hana's sort is documented as being stable, and I would love to steal Brigand's implementation.

**Comment by porkybrain** [comment]
sounds like we need to make it stable

**Comment by ldionne** [comment]
Well, if anything, it would be wiser to provide a `stable_sort` than to mandate the current `sort` to be stable; this way you don't lock yourself into a corner where you _have_ to make your `sort` function stable (as I did).

**Comment by edouarda** [comment]
Got it, thanks.

**Comment by brunocodutra** [comment]
`metal::sort` is stable and, as of brunocodutra/metal@72115c21e0e77380529e463de0a72e2698833e2a, [pretty fast too](https://github.com/edouarda/brigand/issues/194#issuecomment-242886273) ;-)

**Pull request merged**